### ISSUE TITLE
如何为InternVL加入mixture of experts？

### ISSUE THREAD

**Issue by LIKP0** [OpenGVLab/InternVL#535]
您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！

出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）

再次感谢你们的卓越工作！

**Comment by G-z-w** [comment]
感谢，可以参考https://github.com/deepseek-ai/DeepSeek-MoE，希望能对你有帮助。

**Comment by LIKP0** [comment]
谢谢，感谢你的帮助

**Comment by zweiqi** [comment]
> 感谢，可以参考[https://github.com/deepseek-ai/DeepSeek-MoE，希望能对你有帮助。](https://github.com/deepseek-ai/DeepSeek-MoE%EF%BC%8C%E5%B8%8C%E6%9C%9B%E8%83%BD%E5%AF%B9%E4%BD%A0%E6%9C%89%E5%B8%AE%E5%8A%A9%E3%80%82)

hidden_states为视觉特征，为什么经过下面转换后logits变成了nan呢，感觉这个特征不能进行低维度转换
bsz, seq_len, h = hidden_states.shape
--
  | ### compute gating score
  | hidden_states = hidden_states.view(-1, h)
  | logits = F.linear(hidden_states, self.weight, None)

**Comment by zweiqi** [comment]
> 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> 
> 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> 
> 再次感谢你们的卓越工作！

兄弟你实现了吗

**Comment by OuYangg** [comment]
> > 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> > 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> > 再次感谢你们的卓越工作！
> 
> 兄弟你实现了吗

遇到了同样的问题，hidden states后面接个FC就会出现nan，大佬发现原因了嘛

**Comment by zweiqi** [comment]
> > > 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> > > 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> > > 再次感谢你们的卓越工作！
> > 
> > 
> > 兄弟你实现了吗
> 
> 遇到了同样的问题，hidden states后面接个FC就会出现nan，大佬发现原因了嘛

不知道问题在哪唉，无从下手了

**Comment by LIKP0** [comment]
我正在尝试用https://huggingface.co/microsoft/Phi-3.5-MoE-instruct的language model来做MoE，直接替换modeling_intenvl_chat.py里面的Phi3ForCausalLM模型，感觉是种相对简单的方法，还没有细看

**Comment by galactic123** [comment]
> > > > 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> > > > 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> > > > 再次感谢你们的卓越工作！
> > > 
> > > 
> > > 兄弟你实现了吗
> > 
> > 
> > 遇到了同样的问题，hidden states后面接个FC就会出现nan，大佬发现原因了嘛
> 
> 不知道问题在哪唉，无从下手了

同样的问题，我强行用矩阵实现的linear

**Comment by OuYangg** [comment]
> > > > > 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> > > > > 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> > > > > 再次感谢你们的卓越工作！
> > > > 
> > > > 
> > > > 兄弟你实现了吗
> > > 
> > > 
> > > 遇到了同样的问题，hidden states后面接个FC就会出现nan，大佬发现原因了嘛
> > 
> > 
> > 不知道问题在哪唉，无从下手了
> 
> 同样的问题，我强行用矩阵实现的linear

我发现在model = InternVLChatModel.from_pretrained(model_args.model_name_or_path, torch_dtype=torch.bfloat16, config=config)之后再初始化FC就可以解决，否则weight里面会有nan

**Comment by zweiqi** [comment]
> > > > > > 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> > > > > > 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> > > > > > 再次感谢你们的卓越工作！
> > > > > 
> > > > > 
> > > > > 兄弟你实现了吗
> > > > 
> > > > 
> > > > 遇到了同样的问题，hidden states后面接个FC就会出现nan，大佬发现原因了嘛
> > > 
> > > 
> > > 不知道问题在哪唉，无从下手了
> > 
> > 
> > 同样的问题，我强行用矩阵实现的linear
> 
> 我发现在model = InternVLChatModel.from_pretrained(model_args.model_name_or_path, torch_dtype=torch.bfloat16, config=config)之后再初始化FC就可以解决，否则weight里面会有nan

那FC在InternVLChatModel里面的话就不行了吧，属于同时初始化，有点奇怪，为什么同时初始化weight会是nan

**Comment by zweiqi** [comment]
> > > > > > 您好！非常感谢你们的工作，InternVL作为开源模型实在是太酷了，非常有意义的工作！
> > > > > > 出于兴趣，我想探索如何在InternVL中加入混合专家模型，请问有什么建议吗？我的基础知识比较薄弱，不知道该从何处下手，希望有人能给出一些思路，谢谢！我现在使用的是InternVL1.5-Mini4B版本（是的我只能运行这么小的模型...）
> > > > > > 再次感谢你们的卓越工作！
> > > > > 
> > > > > 
> > > > > 兄弟你实现了吗
> > > > 
> > > > 
> > > > 遇到了同样的问题，hidden states后面接个FC就会出现nan，大佬发现原因了嘛
> > > 
> > > 
> > > 不知道问题在哪唉，无从下手了
> > 
> > 
> > 同样的问题，我强行用矩阵实现的linear
> 
> 我发现在model = InternVLChatModel.from_pretrained(model_args.model_name_or_path, torch_dtype=torch.bfloat16, config=config)之后再初始化FC就可以解决，否则weight里面会有nan

请问你怎么初始化的啊，我按照这种先后初始化还是nan
model = InternVLChatModel.from_pretrained(
            model_args.model_name_or_path, torch_dtype=torch.bfloat16, config=config)
vit_hidden_size = config.vision_config.hidden_size
llm_hidden_size = config.llm_config.hidden_size
model.mlp1 =  MLP(num_experts=4, num_selected=2, mm_channels=vit_hidden_size * int(1 / 0.5) ** 2, channels=llm_hidden_size, num_layers=3)

**Comment by OuYangg** [comment]
> InternVLChatModel

我在InternVLChatModel加了个初始化函数

**Comment by zweiqi** [comment]
> > InternVLChatModel
> 
> 我在InternVLChatModel加了个初始化函数

可以分享一下代码吗，本身不都在init初始化函数里吗，没太理解🥹

**Comment by afasijbfk** [comment]
> > > InternVLChatModel
> > 
> > 
> > 我在InternVLChatModel加了个初始化函数
> 
> 可以分享一下代码吗，本身不都在init初始化函数里吗，没太理解🥹
你好，请问您解决这个问题了吗

**Comment by galactic123** [comment]
我解决了

        model = InternVLChatModel.from_pretrained(
            model_args.model_name_or_path, torch_dtype=torch.bfloat16, config=config)
        model.init_extra_weight() # 在modeling_internvl_chat.py 里写个函数重新加载外部权重
